pkg/reconciler/namespacecontroller/namespace_controller.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package namespacecontroller
+
+import (
+	"context"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"kpt.dev/configsync/pkg/status"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// NamespaceController is a controller that watches for Namespace events and
+// sets a flag to indicate whether the Reconciler thread needs to reconcile for
+// this namespace event.
+type NamespaceController struct {
+	client client.Client
+
+	// eventMux is a RMMutex because the cache is read more often than writen.
+	eventMux sync.RWMutex
```

Comment:
I'm not loving how messy this looks with the mutexes and protected values on the controller struct. You also shouldn't have to pass the whole controller to the Run method when it never needs to call Reconcile. That way you dont need to build the whole controller to test the Run func. We probably need some intermediate struct  or something.

Reply:
extracted them into a state structure.